### ISSUE TITLE
Suggestion: allow faker rules to be overridden

### ISSUE THREAD

**Issue by mattwcole** [bchavez/Bogus#42]
I think it would be useful to allow rules to be overridden on fakers. Currently there is an exception thrown when a rule has already been added for the property. If instead, the existing rule was replaced, you could very easily create test data builders like the following.

```csharp
public class FakeCustomerBuilder
{
    private readonly Faker<Customer> _customerFaker;

    public FakeCustomerBuilder()
    {
        _customerFaker = new Faker<Customer>()
            .RuleFor(customer => customer.Forename, faker => faker.Name.FirstName())
            .RuleFor(customer => customer.Surname, faker => faker.Name.LastName());
    }

    public CustomerBuilder WithSurname(string surname)
    {
        _customerFaker.RuleFor(c => c.Surname, faker => surname);
        return this;
    }

    public Customer Build()
    {
        return _customerFaker.Generate();
    }
}
```

**Comment by bchavez** [comment]
@mattwcole No problem, the only problem I see with this approach is if you had multiple specialized methods like `.WithSurname` and `.WithForename`. If this was the case, then you could chainge finalizer actions together, I think with:

```
private Action<Faker, Customer> finalizer = (f, c) => { };
public FakeCustomerBuilder WithSurname(string surname)
{
    this.finalizer += (f, c) => c.Surname = surname;
    return this;
}
public FakeCustomerBuilder WithForename(string forename)
{
    this.finalizer += (f, c) => c.Forename = forename;
    return this;
}
```
Then set `Faker.FinishWith(this.finalizer)` and generate.

Regardless, though, being able to override rules makes sense and probably yield a nicer API to come with what people are expecting, I think. I just need some more time to think about the nature of the issue before loosening up the rules.

Thanks,
Brian

Reference #41, #38, and https://github.com/bchavez/Bogus/issues/31#issuecomment-223148520

**Comment by mattwcole** [comment]
@bchavez you're right, I was just writing a follow up to this. There is also a problem if you want to create a reusable builder. The following will fail because of multiple calls to `FinishWith()`.

```csharp
var builder = new FakeCustomerBuilder();
var mrFoo = builder.WithSurname("foo").Build();
var mrBar = builder.WithSurname("bar").Build();
```

Regardless of the builder pattern, I think in general it could be useful to be able to modify the rules on a `Faker` as needed. Creating a `Faker` with sensible defaults and modifying these later feels like it could be a common use case.

**Comment by bchavez** [comment]
@mattwcole yeah, I think you're right too. I think we've evolved into a good spot now to allow rule overrides given the other pending related issues #41, #38, and #31.

Just trying to reconcile in my mind if there's any value for ***ensuring*** that one, *and only one*, rule has been specified for a property/field on `Faker<T>`; and whether or not it's worth the maintenance effort to keep that existing behavior around through some kind of configuration option.

My feeling is that it's probably not important to ensure that only one rule has been specified and probably not worth the maintenance effort to keep around; since it really is just fake data, not like a domain validation rule. :sunglasses: 

Going to the supermarket :convenience_store: soon and I'll probably have a decision later tonight. If anyone else on the repo watchlist has any additional comments, please feel free to comment.

Thanks,
Brian

:watch: :city_sunset: [***"I just can't wait... I just can't wait... for saturday night..."***](https://www.youtube.com/watch?v=wOjGmpmvGz8#t=2m)

**Comment by bchavez** [comment]
@mattwcole You should have a NuGet package available now [`v8.0.1-beta-1`](https://www.nuget.org/packages/Bogus/8.0.1-beta-1) soon as it gets indexed. Great work m8.

Also, thank you for the unit tests and for of following the repository coding conventions to the `'t'`.

Much respect,
Brian

:car: :blue_car: [***"Let the good times roll..."***](https://www.youtube.com/watch?v=7BDBzgHXf64)
